_Lambda School Data Science, Unit 2_
 
# Sprint Challenge: Predict Steph Curry's shots 🏀

For your Sprint Challenge, you'll use a dataset with all Steph Curry's NBA field goal attempts. (Regular season and playoff games, from October 28, 2009, through June 5, 2019.) 

You'll predict whether each shot was made, using information about the shot and the game. This is hard to predict! Try to get above 60% accuracy. The dataset was collected with the [nba_api](https://github.com/swar/nba_api) Python library.

In [0]:
%%capture
import sys

if 'google.colab' in sys.modules:
    # Install packages in Colab
    !pip install category_encoders==2.*
    !pip install pandas-profiling==2.*

In [1]:
# Read data
import pandas as pd
url = 'https://drive.google.com/uc?export=download&id=1fL7KPyxgGYfQDsuJoBWHIWwCAf-HTFpX'
df = pd.read_csv(url)

# Check data shape
assert df.shape == (13958, 20)

In [93]:
import category_encoders as ce
from sklearn.pipeline import make_pipeline 
from sklearn.impute import SimpleImputer
from sklearn.impute import KNNImputer
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import plot_confusion_matrix
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
from datetime import datetime
import numpy as np
%config IPCompleter.greedy=True

To demonstrate mastery on your Sprint Challenge, do all the required, numbered instructions in this notebook.

To earn a score of "3", also do all the stretch goals.

You are permitted and encouraged to do as much data exploration as you want.

**1. Begin with baselines for classification.** Your target to predict is `shot_made_flag`. What is your baseline accuracy, if you guessed the majority class for every prediction?

**2. Hold out your test set.** Use the 2018-19 season to test. NBA seasons begin in October and end in June. You'll know you've split the data correctly when your test set has 1,709 observations.

**3. Engineer new feature.** Engineer at least **1** new feature, from this list, or your own idea.
- **Homecourt Advantage**: Is the home team (`htm`) the Golden State Warriors (`GSW`) ?
- **Opponent**: Who is the other team playing the Golden State Warriors?
- **Seconds remaining in the period**: Combine minutes remaining with seconds remaining, to get the total number of seconds remaining in the period.
- **Seconds remaining in the game**: Combine period, and seconds remaining in the period, to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.
- **Made previous shot**: Was Steph Curry's previous shot successful?

**4. Decide how to validate** your model. Choose one of the following options. Any of these options are good. You are not graded on which you choose.
- **Train/validate/test split: train on the 2009-10 season through 2016-17 season, validate with the 2017-18 season.** You'll know you've split the data correctly when your train set has 11,081 observations, and your validation set has 1,168 observations.
- **Train/validate/test split: random 80/20%** train/validate split.
- **Cross-validation** with independent test set. You may use any scikit-learn cross-validation method.

**5.** Use a scikit-learn **pipeline** to **encode categoricals** and fit a **Decision Tree** or **Random Forest** model.

**6.** Get your model's **validation accuracy.** (Multiple times if you try multiple iterations.) 

**7.** Get your model's **test accuracy.** (One time, at the end.)


**8.** Given a **confusion matrix** for a hypothetical binary classification model, **calculate accuracy, precision, and recall.**

### Stretch Goals
- Engineer 4+ new features total, either from the list above, or your own ideas.
- Make 2+ visualizations to explore relationships between features and target.
- Optimize 3+ hyperparameters by trying 10+ "candidates" (possible combinations of hyperparameters). You can use `RandomizedSearchCV` or do it manually.
- Get and plot your model's feature importances.



## 1. Begin with baselines for classification. 

>Your target to predict is `shot_made_flag`. What would your baseline accuracy be, if you guessed the majority class for every prediction?

In [3]:
df

game_id  game_event_id    player_name  period  minutes_remaining  \
0      20900015              4  Stephen Curry       1                 11   
1      20900015             17  Stephen Curry       1                  9   
2      20900015             53  Stephen Curry       1                  6   
3      20900015            141  Stephen Curry       2                  9   
4      20900015            249  Stephen Curry       2                  2   
...         ...            ...            ...     ...                ...   
13953  41800403            570  Stephen Curry       4                  8   
13954  41800403            573  Stephen Curry       4                  7   
13955  41800403            602  Stephen Curry       4                  5   
13956  41800403            608  Stephen Curry       4                  4   
13957  41800403            658  Stephen Curry       4                  2   

       seconds_remaining                 action_type       shot_type  \
0                     25                   Jump Shot  3PT Field Goal   
1                     31         Step Back Jump shot  2PT Field Goal   
2                      2                   Jump Shot  2PT Field Goal   
3                     49                   Jump Shot  2PT Field Goal   
4                     19                   Jump Shot  2PT Field Goal   
...                  ...                         ...             ...   
13953                  1            Pullup Jump shot  2PT Field Goal   
13954                 16          Floating Jump shot  2PT Field Goal   
13955                 27         Step Back Jump shot  3PT Field Goal   
13956                 50  Driving Floating Jump Shot  2PT Field Goal   
13957                 47                   Jump Shot  3PT Field Goal   

             shot_zone_basic         shot_zone_area  shot_zone_range  \
0          Above the Break 3  Right Side Center(RC)          24+ ft.   
1                  Mid-Range   Left Side Center(LC)        16-24 ft.   
2      In The Paint (Non-RA)              Center(C)         8-16 ft.   
3                  Mid-Range           Left Side(L)        16-24 ft.   
4                  Mid-Range   Left Side Center(LC)        16-24 ft.   
...                      ...                    ...              ...   
13953  In The Paint (Non-RA)              Center(C)         8-16 ft.   
13954              Mid-Range          Right Side(R)         8-16 ft.   
13955      Above the Break 3   Left Side Center(LC)          24+ ft.   
13956  In The Paint (Non-RA)              Center(C)  Less Than 8 ft.   
13957      Above the Break 3   Left Side Center(LC)          24+ ft.   

       shot_distance  loc_x  loc_y  shot_made_flag   game_date  htm  vtm  \
0                 26     99    249               0  2009-10-28  GSW  HOU   
1                 18   -122    145               1  2009-10-28  GSW  HOU   
2                 14    -60    129               0  2009-10-28  GSW  HOU   
3                 19   -172     82               0  2009-10-28  GSW  HOU   
4                 16    -68    148               0  2009-10-28  GSW  HOU   
...              ...    ...    ...             ...         ...  ...  ...   
13953             12      3    120               1  2019-06-05  GSW  TOR   
13954             11    114     -5               0  2019-06-05  GSW  TOR   
13955             26   -217    149               0  2019-06-05  GSW  TOR   
13956              7     59     49               0  2019-06-05  GSW  TOR   
13957             24   -226    104               0  2019-06-05  GSW  TOR   

          season_type  scoremargin_before_shot  
0      Regular Season                      2.0  
1      Regular Season                      0.0  
2      Regular Season                     -4.0  
3      Regular Season                     -4.0  
4      Regular Season                      0.0  
...               ...                      ...  
13953        Playoffs                    -13.0  
13954        Playoffs                    -14.0  
13955      

In [7]:
df['shot_made_flag'].value_counts(normalize=True)

0    0.527081
1    0.472919
Name: shot_made_flag, dtype: float64

Baseline majority class would give you 52.7% accuracy.

## 2. Hold out your test set.

>Use the 2018-19 season to test. NBA seasons begin in October and end in June. You'll know you've split the data correctly when your test set has 1,709 observations.

In [8]:
df.dtypes

game_id                      int64
game_event_id                int64
player_name                 object
period                       int64
minutes_remaining            int64
seconds_remaining            int64
action_type                 object
shot_type                   object
shot_zone_basic             object
shot_zone_area              object
shot_zone_range             object
shot_distance                int64
loc_x                        int64
loc_y                        int64
shot_made_flag               int64
game_date                   object
htm                         object
vtm                         object
season_type                 object
scoremargin_before_shot    float64
dtype: object

In [19]:
# need to convert game_date
df['game_date'] = pd.to_datetime(df['game_date'])

test = df[(df['game_date'] >= '10-01-2018') & (df['game_date'] < '07-01-2019')]
test.shape

(1709, 20)

In [74]:
train = df[df['game_date'] < '10-01-2018']
train.isnull().sum()

game_id                    0
game_event_id              0
player_name                0
period                     0
minutes_remaining          0
seconds_remaining          0
action_type                0
shot_type                  0
shot_zone_basic            0
shot_zone_area             0
shot_zone_range            0
shot_distance              0
loc_x                      0
loc_y                      0
shot_made_flag             0
game_date                  0
htm                        0
vtm                        0
season_type                0
scoremargin_before_shot    0
time_left_in_per           0
time_left_in_game          0
sec_left_in_per            0
sec_left_in_game           0
dtype: int64

## 3. Engineer new feature.

>Engineer at least **1** new feature, from this list, or your own idea.
>
>- **Homecourt Advantage**: Is the home team (`htm`) the Golden State Warriors (`GSW`) ?
>- **Opponent**: Who is the other team playing the Golden State Warriors?
>- **Seconds remaining in the period**: Combine minutes remaining with seconds remaining, to get the total number of seconds remaining in the period.
>- **Seconds remaining in the game**: Combine period, and seconds remaining in the period, to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.
>- **Made previous shot**: Was Steph Curry's previous shot successful?

    

In [35]:
len(df['htm'])

13958

In [54]:
# Will come back to this attempt. for now next feature engineered
# home_away = []
# def opp(df):
#     for i in range(0,len(df['htm'])):
#         if df['htm'][i] == 'GSW':
#             home_away.append(df['vtm'][i])
#         else:
#             home_away.append(df['htm'][i])


In [55]:
# df1 = df.apply(opp, axis = 1)

In [57]:
# len(home_away)

41874

In [68]:
# Seconds remaining in the game: Combine period, and seconds remaining in the period, 
# to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.

# well we  should actually use minutes remaining as well

df['sec_left_in_per'] = df['minutes_remaining']*60 + df['seconds_remaining']
df['sec_left_in_game'] = (4-df['period'])*12*60 + df['sec_left_in_per']
df['sec_left_in_game']

0        2845
1        2731
2        2522
3        2029
4        1579
         ... 
13953     481
13954     436
13955     327
13956     290
13957     167
Name: sec_left_in_game, Length: 13958, dtype: int64

In [72]:
df1 = df.drop(columns=['sec_left_in_per', 'time_left_in_game', 'time_left_in_per'])

In [99]:
# Let's get the train and test split again
test1 = df1[(df1['game_date'] >= '10-01-2018') & (df1['game_date'] < '07-01-2019')]
print('test1 shape:', test1.shape)
train1 = df1[df1['game_date'] < '10-01-2018']
train1 = train1.drop(columns='game_date')

test1 shape: (1709, 21)


## **4. Decide how to validate** your model. 

>Choose one of the following options. Any of these options are good. You are not graded on which you choose.
>
>- **Train/validate/test split: train on the 2009-10 season through 2016-17 season, validate with the 2017-18 season.** You'll know you've split the data correctly when your train set has 11,081 observations, and your validation set has 1,168 observations.
>- **Train/validate/test split: random 80/20%** train/validate split.
>- **Cross-validation** with independent test set. You may use any scikit-learn cross-validation method.

Im going with the cross-val method

## 5. Use a scikit-learn pipeline to encode categoricals and fit a Decision Tree or Random Forest model.

In [102]:
target = 'shot_made_flag'
features = train1.columns
features = features.drop('shot_made_flag')

In [103]:
X_train = train1[features]
y_train = train1[target]
X_test = test1[features]
y_test = test1[target]

In [97]:
type(X_train)

pandas.core.frame.DataFrame

In [126]:
pipeline = make_pipeline(
    ce.OrdinalEncoder(),
    RandomForestClassifier(random_state=15)
)
pipeline.fit(X_train, y_train)

Pipeline(memory=None,
         steps=[('ordinalencoder',
                 OrdinalEncoder(cols=['player_name', 'action_type', 'shot_type',
                                      'shot_zone_basic', 'shot_zone_area',
                                      'shot_zone_range', 'htm', 'vtm',
                                      'season_type'],
                                drop_invariant=False, handle_missing='value',
                                handle_unknown='value',
                                mapping=[{'col': 'player_name',
                                          'data_type': dtype('O'),
                                          'mapping': Stephen Curry    1
NaN             -2
dtype: int64},
                                         {'...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                   

## 6.Get your model's validation accuracy

> (Multiple times if you try multiple iterations.)

In [129]:
k=3
scores = cross_val_score(pipeline, X_train, y_train, cv=k, 
                         scoring='accuracy')
print(f' Accuracy for {k} folds:', scores)
print('Mean', scores.mean())

 Accuracy for 3 folds: [0.58584374 0.62405094 0.63090865]
Mean 0.613601110294718


## 7. Get your model's test accuracy

> (One time, at the end.)

In [130]:
y_pred = pipeline.predict(X_test)
print('Test Accuracy:', accuracy_score(y_test, y_pred))

Test Accuracy: 0.6184903452311293


## 8. Given a confusion matrix, calculate accuracy, precision, and recall.

Imagine this is the confusion matrix for a binary classification model. Use the confusion matrix to calculate the model's accuracy, precision, and recall.

<table>
  <tr>
    <td colspan="2" rowspan="2"></td>
    <td colspan="2">Predicted</td>
  </tr>
  <tr>
    <td>Negative</td>
    <td>Positive</td>
  </tr>
  <tr>
    <td rowspan="2">Actual</td>
    <td>Negative</td>
    <td style="border: solid">85</td>
    <td style="border: solid">58</td>
  </tr>
  <tr>
    <td>Positive</td>
    <td style="border: solid">8</td>
    <td style="border: solid"> 36</td>
  </tr>
</table>

### Calculate accuracy 

In [131]:
correct_pred = 85 + 36
total_pred = 85 + 58 + 8 + 36
class_act = correct_pred /  total_pred
print('Accuracy:', class_act)

Accuracy: 0.6470588235294118


### Calculate precision

In [134]:
# of all you predicted, how many right

# For negative
correct_pred_neg = 85
total_pred_neg = 85 + 8
Precision_neg = correct_pred_neg/total_pred_neg
print('Precision for negative:', Precision_neg)

# For positive just to check
correct_pred_pos = 36
total_pred_pos = 36 + 58
Precision_pos = correct_pred_pos/total_pred_pos
print('Precision for positive:', Precision_pos)

Precision for negative: 0.9139784946236559
Precision for positive: 0.3829787234042553


### Calculate recall

In [136]:
# of all the actual instances of the class, how many predicted correctly?

# For neg
actual_neg = 85 + 58
neg_recall = correct_pred_neg / actual_neg
print('Negative recall:', neg_recall)

# For pos
actual_pos = 8 + 36
pos_recall = correct_pred_pos / actual_pos
print('Positive recall:', pos_recall)


Negative recall: 0.5944055944055944
Positive recall: 0.8181818181818182
